src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,550 @@
+        {
+            Requires.NotNull(moniker, nameof(moniker));
+
+            await InitializeAsync();
+
+            await UnconfiguredProjectServices.ThreadingService.SwitchToUIThread();
+
+            DesignTimeInputsItem inputs = AppliedValue.Value;
+
+            if (!inputs.Inputs.Contains(moniker))
+                throw new ArgumentException("Moniker supplied must be one of the DesignTime source files", nameof(moniker));
+
+            string outputPath = inputs.OutputPath;
+
+            HashSet<string> files = GetFilesToCompile(moniker, inputs.SharedInputs);
+            string outputFileName = GetOutputFileName(outputPath, moniker);
+            if (CompilationNeeded(files, outputFileName))
+            {
+                if (inputs.TaskSchedulers.TryGetValue(moniker, out ITaskDelayScheduler scheduler))
+                {
+                    // For parity with legacy we don't care about the compilation result: Legacy only errors here if it runs out of memory queuing the compilation
+                    // Additionally for parity, we compile here on the UI thread and block (whilst still preventing simultaneous work)
+                    await scheduler.RunAsyncTask(token => CompileTempPEAsync(files, outputFileName, token));
+                }
+            }
+
+            // VSTypeResolutionService is the only consumer, and it only uses the codebase element so it's fine to default most of these (VC++ does the same)
+            return $@"<root>
+  <Application private_binpath = ""{outputPath}""/>
+  <Assembly
+    codebase = ""{outputFileName}""
+    name = ""{moniker}""
+    version = ""0.0.0.0""
+    snapshot_id = ""1""
+    replaceable = ""True""
+  />
+</root>";
+        }
+
+        internal bool CompilationNeeded(HashSet<string> files, string outputFileName)
+        {
+            if (!_fileSystem.FileExists(outputFileName))
+                return true;
+
+            try
+            {
+                DateTime outputDateTime = _fileSystem.LastFileWriteTimeUtc(outputFileName);
+
+                foreach (string file in files)
+                {
+                    DateTime fileDateTime = _fileSystem.LastFileWriteTimeUtc(file);
+                    if (fileDateTime > outputDateTime)
+                        return true;
+                }
+            }
+            // if we can't read the file time of the output file, then we presumably can't compile to it either, so returning false is appropriate.
+            // if we can't read the file time of an input file, then we presumably can't read from it to compile either, so returning false is appropriate
+            catch (IOException)
+            { }
+            catch (UnauthorizedAccessException)
+            { }
+
+            return false;
+        }
+
+        private static string GetOutputFileName(string outputPath, string moniker)
+        {
+            // All monikers are project relative paths by defintion (anything else is a link), meaning the only invalid
```

Comment:
I don't under the why about "anything else is a link" it seems like there's an assumption here not called out, are linked files excluded?

Reply:
Yes, linked files aren't TempPE inputs. It's called out in a comment below, but I'll make it clearer.
https://github.com/dotnet/project-system/pull/4740/files#diff-8cf178a6054d65774e94e9f40a9a353eR492